File: ios/Runner/Channels/CustomConfig/CustomConfigChannel.swift
//
//  CustomConfigChannel.swift
//  Runner
//

import Foundation
import UIKit
import Flutter
import VGSCheckoutSDK

/// Implementation of Flutter Bridge (FlutterMethodChannel).
class CustomConfigChannel: NSObject {

	// MARK: - Initialization.

	/// Initializer.
	/// - Parameter messenger: `FlutterBinaryMessenger` object, binary messenger,
	init(messenger: FlutterBinaryMessenger) {
		// Create Flutter method channel to invoce methods from Flutter code in native iOS code by sending messages.
		flutterMethodChannel = FlutterMethodChannel(name: "vgs.com.checkout/customConfig",
																							binaryMessenger: messenger)
		super.init()
		registerCustomConfigChannel()
	}

	// MARK: - Vars

	/// Checkout instance.
	fileprivate var vgsCheckout: VGSCheckout?

	/// Flutter method channel.
	fileprivate var flutterMethodChannel: FlutterMethodChannel

	// MARK: - Method Channel

	/// Registers custom config channel.
	/// - Parameter controller: `FlutterViewController` object, Flutter view controller.
	func registerCustomConfigChannel() {

		// Set handler to invoce native iOS code on messages from Flutter code.
		flutterMethodChannel.setMethodCallHandler({[weak self]
			(call: FlutterMethodCall, result: @escaping FlutterResult) -> Void in
			// This method is invoked on the UI thread.
			guard call.method == "startCustomCheckoutConfig" else {
				result(FlutterMethodNotImplemented)
				return
			}

			self?.startCustomCheckoutConfig(result: result)
		})
	}

	/// Start custom checkout configuration.
	/// - Parameter result: `FlutterResult` object, flutter result.
	private func startCustomCheckoutConfig(result: @escaping FlutterResult) {

		// Use root view controller (acts like a main app widget) to present checkout from.
		let controller : FlutterViewController = UIApplication.shared.keyWindow!.rootViewController as! FlutterViewController

		var checkoutConfiguration = VGSCheckoutCustomConfiguration(vaultID: DemoAppConfiguration.shared.vaultId, environment: DemoAppConfiguration.shared.environment)

		checkoutConfiguration.cardHolderFieldOptions.fieldNameType = .single("cardHolder_name")
		checkoutConfiguration.cardNumberFieldOptions.fieldName = "card_number"
		checkoutConfiguration.expirationDateFieldOptions.fieldName = "exp_data"
		checkoutConfiguration.cvcFieldOptions.fieldName = "card_cvc"

		checkoutConfiguration.billingAddressVisibility = .visible

		checkoutConfiguration.billingAddressCountryFieldOptions.fieldName = "billing_address.country"
		checkoutConfiguration.billingAddressCityFieldOptions.fieldName = "billing_address.city"
		checkoutConfiguration.billingAddressLine1FieldOptions.fieldName = "billing_address.addressLine1"
		checkoutConfiguration.billingAddressLine2FieldOptions.fieldName = "billing_address.addressLine2"
		checkoutConfiguration.billingAddressPostalCodeFieldOptions.fieldName = "billing_address.postal_code"

		// Produce nested json for fields with `.` notation.
		checkoutConfiguration.routeConfiguration.requestOptions.mergePolicy = .nestedJSON

		checkoutConfiguration.routeConfiguration.path = "post"

		/* Set custom date user input/output JSON format.

		checkoutConfiguration.expirationDateFieldOptions.inputDateFormat = .shortYearThenMonth
		checkoutConfiguration.expirationDateFieldOptions.outputDateFormat = .longYearThenMonth

		let expDateSerializer = VGSCheckoutExpDateSeparateSerializer(monthFieldName: "card_date.month", yearFieldName: "card_date.year")
		checkoutConfiguration.expirationDateFieldOptions.serializers = [expDateSerializer]
		*/

		// Init Checkout with vault and ID.
		vgsCheckout = VGSCheckout(configuration: checkoutConfiguration)

		//VGSPaymentCards.visa.formatPattern = "#### #### #### ####"

		/// Change default valid card number lengthes
//		VGSPaymentCards.visa.cardNumberLengths = [16]
//		/// Change default format pattern
//		VGSPaymentCards.visa.formatPattern = "#### #### #### ####"

		// Listen to events from Checkout in native iOS code.
		// Delegate is native iOS pattern which implies emitter and single subscriber linked by each other with weak reference semantic.
		vgsCheckout?.delegate = self

		// Present checkout configuration.
		vgsCheckout?.present(from: controller)

		// Call result object after presenting checkout with 0.3s delay.
		DispatchQueue.main.asyncAfter(deadline: .now() + 0.3) {
			result(nil)
		}
	}
}

// MARK: - VGSCheckoutDelegate

extension CustomConfigChannel: VGSCheckoutDelegate {
	func checkoutDidCancel() {
		flutterMethodChannel.invokeMethod("handleCancelCheckout", arguments: nil)
	}

	func checkoutDidFinish(with requestResult: VGSCheckoutRequestResult) {

		var title = ""
		var message = ""

		switch requestResult {
		case .success(let statusCode, let data, let response, let info):
			title = "Checkout status: Success!"
			message = "status code is: \(statusCode)"
			let text = DemoAppResponseParser.stringifySuccessResponse(from: data) ?? ""

			var payload = [String: Any]()
			payload["STATUS"] = "FINISHED_SUCCESS"
			payload["DESCRIPTION"] = text
			payload["DATA"] = DemoAppResponseParser.convertToJSON(from: data)

			flutterMethodChannel.invokeMethod("handleCheckoutSuccess", arguments: payload)
		case .failure(let statusCode, let data, let response, let error, let info):
			title = "Checkout status: Failed!"
			message = "status code is: \(statusCode) error: \(error?.localizedDescription ?? "Uknown error!")"

			var payload = [String: Any]()
			payload["STATUS"] = "FINISHED_ERROR"
			payload["DATA"] = message
			flutterMethodChannel.invokeMethod("handleCheckoutFail", arguments: payload)
		}
	}
}

File: android/app/src/main/kotlin/com/example/vgs_checkout_flutter_demo/channel/CustomConfigChannel.kt
package com.example.vgs_checkout_flutter_demo.channel

import android.util.Log
import androidx.fragment.app.Fragment
import com.example.vgs_checkout_flutter_demo.utils.toEnvironment
import com.google.gson.Gson
import com.verygoodsecurity.vgscheckout.VGSCheckout
import com.verygoodsecurity.vgscheckout.VGSCheckoutCallback
import com.verygoodsecurity.vgscheckout.config.VGSCheckoutCustomConfig
import com.verygoodsecurity.vgscheckout.config.networking.request.core.VGSCheckoutDataMergePolicy
import com.verygoodsecurity.vgscheckout.config.ui.view.address.VGSCheckoutBillingAddressVisibility
import com.verygoodsecurity.vgscheckout.model.VGSCheckoutResult
import com.verygoodsecurity.vgscheckout.model.VGSCheckoutResult.*
import com.verygoodsecurity.vgscheckout.model.VGSCheckoutResultBundle.Keys.ADD_CARD_RESPONSE
import com.verygoodsecurity.vgscheckout.model.response.VGSCheckoutCardResponse
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.MethodChannel

class CustomConfigChannel(fragment: Fragment, engine: FlutterEngine) : VGSCheckoutCallback {

    private var checkout: VGSCheckout
    private var channel: MethodChannel

    init {
        checkout = VGSCheckout(fragment, this)
        channel = MethodChannel(engine.dartExecutor.binaryMessenger, CHANNEL_NAME)
        channel.setMethodCallHandler { call, result ->
            val arguments = call.arguments as HashMap<*, *>
            when (call.method) {
                METHOD_NAME -> startCustomCheckoutConfig(
                    arguments["vault_id"] as String,
                    arguments["environment"] as String
                )
                else -> result.notImplemented()
            }
        }
    }

    override fun onCheckoutResult(result: VGSCheckoutResult) {
        val data = getResultData(result)
        when (result) {
            is Success -> channel.invokeMethod("handleCheckoutSuccess", data)
            is Failed -> channel.invokeMethod("handleCheckoutFail", data)
            is Canceled -> channel.invokeMethod("handleCancelCheckout", null)
        }
    }

    private fun startCustomCheckoutConfig(vaultId: String, environment: String) {
        Log.d("CustomConfigChannel", "vaultId = $vaultId")
        Log.d("CustomConfigChannel", "environment = $environment")
        val config = createConfig(vaultId, environment)
        checkout.present(config)
    }

    private fun createConfig(vaultId: String, environment: String): VGSCheckoutCustomConfig {
        return VGSCheckoutCustomConfig.Builder(vaultId)
            .setEnvironment(environment.toEnvironment())
            .setCardHolderOptions("card_holder_name")
            .setCardNumberOptions("card_number")
            .setExpirationDateOptions("exp_data")
            .setCVCOptions("card_cvc")
            .setBillingAddressVisibility(VGSCheckoutBillingAddressVisibility.VISIBLE)
            .setCountryOptions("billing_address.country")
            .setCityOptions("billing_address.city")
            .setAddressOptions("billing_address.addressLine1")
            .setOptionalAddressOptions("billing_address.addressLine2")
            .setPostalCodeOptions("billing_address.postal_code")
            .setMergePolicy(VGSCheckoutDataMergePolicy.NESTED_JSON)
            .setPath("post")
            .build()
    }

    private fun getResultData(result: VGSCheckoutResult): Map<String, Any> {
        val data = mutableMapOf<String, Any>()
        val response = result.data.getParcelable<VGSCheckoutCardResponse>(ADD_CARD_RESPONSE)?.body
        data["STATUS"] = if (result is Success) "FINISHED_SUCCESS" else "FINISHED_ERROR"
        if (!response.isNullOrEmpty()) data["DATA"] = Gson().fromJson(response, Map::class.java)
        data["DESCRIPTION"] = "Success: \n $response"
        return data
    }

    private companion object {

        const val CHANNEL_NAME = "vgs.com.checkout/customConfig"
        const val METHOD_NAME = "startCustomCheckoutConfig"
    }
}